### PR TITLE
refactor: traces and dijkstra algorithm

### LINKED ISSUE #76: optimize priority heap in Dijkstra implementation
Currently we use the Java's default PriorityQueue which, although quite fast, is the bottleneck of the Dijkstra algorithm.

There are several priority queue implementations based on different techniques. A known alternative are fibonacci heaps. There is a mutable [Java implementation](https://github.com/jgrapht/jgrapht/blob/master/jgrapht-core/src/main/java/org/jgrapht/util/FibonacciHeap.java) and an immutable Clojure implementation but according to its author it seems to be [quite slow](https://github.com/maryrosecook/dijkstra) ([blog post](https://maryrosecook.com/blog/post/the-fibonacci-heap-ruins-my-life)). 

Both of these algorithms are theoretically sound but according to [this blog post](http://queue.acm.org/detail.cfm?id=1814327) it seems that they are not that good in practice. The author proposes a [B-heap](https://www.wikiwand.com/en/B-heap) as a solution.

[This](https://github.com/varnish/Varnish-Cache/blob/master/lib/libvarnish/binary_heap.c) C implementation might serve as blueprint for a Java one. Although I am not sure of the applicability of this technique to Java since it is a memory-managed language

extra readings:
- [priority queue](https://www.wikiwand.com/en/Priority_queue)
- [benchmark of different Fibonacci and binary heaps](https://gabormakrai.wordpress.com/2015/02/11/experimenting-with-dijkstras-algorithm/)
- [cache oblivious structures](https://blogs.msdn.microsoft.com/devdev/2007/06/12/cache-oblivious-data-structures/) with shortest path, and priority queue recommendations
- [cache oblivious structures paper](https://pdfs.semanticscholar.org/01f8/dc5bcb94b52a92237de3f8c527d8f5c60c6f.pdf)

TODO:
- [x] measure and publish the current CPU profiling of a Dijkstra execution
- [x] test and benchmark priority queues implementation (in the context of Dijkstra executions !!)
- ~[ ] test and benchmark using an [AVL set](https://github.com/clojure/data.avl/) as a priority queue~

### LINKED ISSUE #80: replace Trace by List and Map$Entry
The Trace implementation convolutes a sequence and a key-value pair. This makes it both difficult to understand and curversome to use.

A better approach, although maybe slower?, would be to use a Clojure List as container and a Map$Entry as elements.

### LINKED ISSUE #81: consider using a HashMap for reverse mapping in Dijkstra
This would avoid the use of `settled` and enable the use of `decreaseKey`

example: http://www.keithschwarz.com/interesting/code/?dir=dijkstra

### LINKED ISSUE #83: create a Graph type
Although the current IntMap works well, it has several problem when using dissoc or similar functions which require some book keeping

### PR DESCRIPTION
**Pull request by carocad**
This is a performance optimization. The new code is *8 times faster* than the old one :)

fixes #76 - using Fibonnaci heaps
fixes #80 - using Java's SimpleImmutableMapEntry 
fixes #83 - `detach` function is enough for most practical purposes
fixes #81 - use a simple mutable hashmap inside djijkstra with a lazy (immutable) sequence exposed to the outside

### CODE REVIEW

**Comment by carocad** [comment]
@mehdisadeghi sorry for such a massive PR. 

After improving the performance of dijkstra lots of things started to fall apart so I had to create more test and fix those things as well. So the small optimization turn out to be a big refactor job. The linked tickets and the tests added should explain what I did though ;)

**Comment by carocad** [comment]
PS: I havent tested the returned JSON yet. My guess is that there might be problems there but for the time being I didnt want to make the PR even bigger

File: resources/benchmark.txt [hunk 1]
```diff
@@ -6,72 +6,85 @@ DIJKSTRA forward with: 90397 nodes and 230142 edges
 saarland graph:
 amd64 Linux 4.4.0-97-generic 4 cpu(s)
 OpenJDK 64-Bit Server VM 25.131-b11
-Runtime arguments: -Dfile.encoding=UTF-8 -Xmx1g -XX:-OmitStackTraceInFastThrow -Dclojure.compile.path=/home/carocad/Proyectos/kamal/target/classes -Dkamal.version=0.1.4 -Dclojure.debug=false
+Runtime arguments: -Dfile.encoding=UTF-8 -Xmx1g -XX:-OmitStackTraceInFastThrow -Dclojure.compile.path=/home/carocad/Proyectos/kamal/target/classes -Dkamal.version=0.2.0 -Dclojure.debug=false
```
Comment (mehdisadeghi) [inline resources/benchmark.txt]: Could you add a short guide to the wiki about benchmarks?

Comment (carocad) [reply resources/benchmark.txt]: can you tell me what would you expect to see there? I guess I am too deep in the topic so I think that most of the stuff should be clear but obviously that is never the case :rofl:

Comment (mehdisadeghi) [reply resources/benchmark.txt]: Just basic stuff that allow people know which tools and commands they should use and how to update the files. A quick review of the purpose of the files would be very useful too.

**Pull request merged**